README.md

```diff
@@ -145,6 +145,20 @@ Notes when working with Binaryen IR:
    incorrectly.
  * For similar reasons, nodes should not appear in more than one functions.
 
+Notes on trap semantics and optimization:
+
+ * Binaryen's optimization passes generally do not change program semantics, but
+   an exception is made for traps. In particular, optimizations may change
+   operations that might implicitly trap to operations that will never trap but
+   are otherwise equivalent. For example, a trapping float-to-int conversion may
+   be changed to a nontrapping float-to-int conversion.
+ * By default, Binaryen's optimizations will assume that trapping instructions
+   may trap unless they can prove otherwise. This conservative assumption can
+   inhibit some optimizations, so users can tell the optimizer to assume that
+   implicit traps will never occur by passing `--ignore-implicit-traps`. This
+   option may cause output programs to trap when the input program would not
+   have trapped, so it should be used with caution.
```

Comment:
So does this mean
- Originally trapping program may not trap, regardless of options
- Originally non-trapping program may trap, if you give `--ignore-implicit-traps`
?

- Why does ignoring implicit traps cause traps to otherwise non-trapping code?
- In which case Binaryen remove traps? Memory packing?

Reply:
> So does this mean
> 
> 1. Originally trapping program may not trap, regardless of options

Yes, but only for implicit traps. `unreachable` is an explicit trap and would never be removed.

> 2. Originally non-trapping program may trap, if you give `--ignore-implicit-traps`?

Yep

> * Why does ignoring implicit traps cause traps to otherwise non-trapping code?

For example, if implicit traps are ignored, the optimizer can assume that a conditionally-executed trapping float-to-int conversion will never trap and is therefore safe to execute unconditionally instead. But that condition might have prevented the conversion from being executed in cases that would have caused it to trap and those cases may come up in real execution.

> * In which case Binaryen remove traps? Memory packing?

Yes, that's the only one I can think of right now.